### PR TITLE
Validate collections yaml

### PR DESCRIPTION
**Pull request by katjuell**
### What should this PR do?
Resolves [DEVED-22](https://sourcegraph.atlassian.net/browse/DEVED-22) by adding validation for our `collections.yaml`.

Refactors `loadCollections` as part of this, also with a future eye to enabling content type checks as part of the validation flow (i.e. if no collection of `type = recordType` then skip).

Also prefers `null` to undefined and setting empty strings as default values.

### Why are we making this change?
We would like to be able to validate our data, ensuring it matches the intended format.

### What are the acceptance criteria? 
- Validations should work for record collections.
- Validations should work for author collections.
- Site should work as expected.

### How should this PR be tested?
- Check out branch, ensure site is working as expected.
- Ensure no new errors on linting.

## Pull request process

**Reviewers**:

1. Test functionality using the criteria above.
2. Offer tips for efficiency, feedback on best practices, and possible alternative approaches and things that may not have been considered.
3. For shorter, "quick" PRs, use your best judgement on #​2.
4. Use a collaborative approach and provide resources and/or context where appropriate.
5. Provide screenshots/grabs where appropriate to show findings during review.

**Reviewees**:

1. Prefer incremental and appropriately-scoped changes.
2. Leave a comment on things you want explicit feedback on.
3. Respond clearly to comments and questions.

### CODE REVIEW

File: collections.yaml [hunk 1]
```diff
@@ -1,5 +1,6 @@
 collections:
     - title: How To Use Search Patterns on Sourcegraph
+      slug: how-to-use-search-patterns-on-sourcegraph
```
Comment (katjuell) [inline collections.yaml:3]: I've changed this to be required as I think it could be useful to have down the road.

File: components/Article.tsx [hunk 2]
```diff
@@ -18,17 +18,17 @@ import TocWrapper from './TocWrapper'
 
 export interface Props {
     title: string
-    author: string
+    author?: string | null
```
Comment (katjuell) [inline components/Article.tsx:21]: We have this marked as optional in `FrontMatter` so updated this to match.

File: components/AuthorCard.tsx [hunk 3]
```diff
@@ -1,11 +1,9 @@
 import Link from 'next/link'
 import React from 'react'
 
-import { AuthorDefinition } from '../util/loadCollections'
-
 interface Props {
     name: string
-    bio: string
+    bio?: string
```
Comment (katjuell) [inline components/AuthorCard.tsx]: Seems like we are treating this as optional for now?

File: util/loadCollections.ts [hunk 4]
```diff
@@ -2,59 +2,80 @@ import { promises as fs } from 'fs'
...
+    let uniqueMemberRecords: MarkdownFile[] = []
+    collectionDefinition.members.map(memberSlug => {
+        const memberRecord = findMemberRecord(records, memberSlug, collectionDefinition)
+        uniqueMemberRecords = [ ...uniqueMemberRecords, memberRecord ]
+    })
+
+    return uniqueMemberRecords
+}
+
+function returnRecordCollections(collections: RecordCollectionDefinition[], records: MarkdownFile[]): RecordCollection[] {
```
Comment (katjuell) [inline util/loadCollections.ts:43]: Refactoring things this way made sense to me for two reasons: 1. It enabled the yaml validation in a place that I thought made sense (before we fetch records); and 2. I thought it would be easier down the road to embed a record type check here.

Breaking these loops into more methods also means more code (le sigh), but hopefully there is a readability boost?

**Comment by marekweb** [comment]
Nice! Before I comment on the code itself I wanted to add some background context because the collections code wasn't in an entirely finished state and we hadn't settled on naming.

In terms of naming/vocabulary, I intended the name "collections" to refer to a "collection of posts" and as a synonym of "series", when I worked on #103. I wanted a name that was more general than "series" at the time. So `collections.yaml` started out as a file containing those collections.

Then we wanted to add "authors" as a new yaml data type and we took a shortcut (it was during a pair-progrmaming session and we wanted a quick prototype) and added it directly into `collections.yaml` with the intent of renaming that file to something else like `data.yaml`. We also considered splitting it into two files in a data directory: `data/authors.yaml` and `data/collections.yaml` which would allow for a similar pattern for other yaml data types in the future. In the meantime we may also have used `collections` in a generic way to mean `a collection of any record` (in the style of the vocabulary that Mongodb uses for example) which I think just added to the confusion.

In conclusion the current naming was experimental and it would be good to settle on clear and consistent names going forward, which is why I'm bringing this up on this PR. We considered using the word `series` instead of calling them `collections` but the downside is that `series` is both a singular and a plural, and I feel that this makes it a potentially unclear name for variables and function names.

**Comment by katjuell** [comment]
This is super helpful — thank you! I love the idea of having a `data` folder with more specifically-scoped files. I think `collections` makes sense (and I agree on singular v plural!), so what if I refactor this to see what `data/collections.yaml` and `data/authors.yaml` would look like?

File: util/loadCollections.ts (hunk 4 above)
Comment (ltagliaferri) [reply util/loadCollections.ts:37]: +1 readability

File: package.json [hunk 5]
```diff
@@ -10,6 +10,8 @@
     "prettier": "prettier '**/*.{js?(on),ts?(x),md}' --ignore-path .gitignore --write --list-different",
     "eslint": "eslint '**/*.{ts,tsx}' --ignore-path .gitignore",
     "eslint-fix": "npm run eslint -- --fix",
+    "check-types": "tsc --noemit",
```
Comment (katjuell) [inline package.json:13]: Sneaking this in here mainly to help myself, but I can move to a separate PR if folks feel strongly about the scope creep.

**Comment by ltagliaferri** [comment]
I like the `data/collections.yaml` and `data/authors.yaml` solution. I think "collections" could be a parent to "series" and also "posts" ("resources" is what I am trying to socialize as more esteemed than "post" 😄 , e.g. "educational resource" which could be a catchall of "tutorial" / "video" / "conceptual article" / etc).

File: util/AuthorCollectionDefinition.ts [hunk 6]
```diff
@@ -0,0 +1,10 @@
+/**
```
Comment (katjuell) [inline util/AuthorCollectionDefinition.ts:1]: Going to address where these live in [DEVED-74](https://sourcegraph.atlassian.net/browse/DEVED-74)

**Comment by marekweb** [comment]
> "resources" is what I am trying to socialize as more esteemed than "post" , e.g. "educational resource" 

:+1: to this, "resource" is a much better name for the concept of "a piece of content" than "post"

File: data/authors.yaml [hunk 7]
```diff
@@ -15,6 +7,3 @@ authors:
         - https://www.linkedin.com/in/marekzaluski
     - id: lisa-tagliaferri
       name: Lisa Tagliaferri
-      bio: TODO
```
Comment (katjuell) [inline data/authors.yaml:18]: I took these placeholders out — to me, not present looks better than `todo`

File: components/Article.tsx [hunk 8]
```diff
@@ -18,17 +18,17 @@ import TocWrapper from './TocWrapper'
...
-    socialImage?: string
-    description?: string
-    toc?: Node
-    collection?: RecordCollection
+    image?: string | null
+    imageAlt?: string | null
+    socialImage?: string | null
+    description?: string | null
+    toc?: Node | null
+    collection?: RecordCollection | null
```
Comment (marekweb) [inline components/Article.tsx:29]: I'm curious about the reason for the `null` in these fields. I would assume that it's simpler to only allow `undefined`, which we already have with the optional nature of the fields, and that covers the case of the field being entirely absent from the object as well as explicitly having an `undefined` value.

File: package.json (hunk 5 above)
Comment (marekweb) [reply package.json:13]: :+1: - this is a great idea.

File: components/Article.tsx (hunk 8 above)
Comment (katjuell) [reply components/Article.tsx:29]: I was thinking about it for our validators — where we are basically trying to "normalize" our data.  I didn't like the idea of setting empty strings or other placeholder values that don't themselves have meaning, and may be hard to reason about in the future, and I thought `null` might be the way to go — thinking down the road as well for interfacing with a database or some such (reasoning about `NULL` versus empty strings or placeholders). If we were using PostgreSQL, for example, the default on a value that's not set would be `NULL` (unless another default is set).

**Comment by marekweb** [comment]
Do we want to go ahead with picking a consistent meaning for "collections" in this PR or leave it for another PR?

If yes then it looks like it's a question of choosing between:

1. "collection" meaning a list of objects (in the sense of "collection of database records")
2. "collection" meaning a series of resources (in the sense of "This article is part of a series...")

I think choice 2 is clearer because we can just use plurals to mean "a list of...". In that vein I think functions names like `loadAuthors` and `loadCollections` will be clearer than `loadAuthorCollections` and `loadRecordCollections`.

File: util/omitUndefinedFields.ts [hunk 9]
```diff
@@ -13,7 +13,7 @@ export default function omitUndefinedFields<T>(object: T): T {
     }
     return Object.fromEntries(
         Object.entries(object)
-            .filter(([, value]) => value !== undefined)
+            .filter(([, value]) => value !== null)
             .map(([key, value]) => {
```
Comment (marekweb) [inline util/omitUndefinedFields.ts:17]: Did you encounter an issue with `null` in `omitUndefinedFields`? Here, we need to filter `undefined` here because Next.js has an issue with it, but it doesn't have an issue with `null`. I believe there's a bug in `omitUndefinedFields` because `typeof null` evaluates to `'object'` but the above check doesn't handle that. I think if we fix that bug, then we can keep this `filter` line unchanged.

**Comment by katjuell** [comment]
I think maybe I'm a bit unclear on option 2, so let me step through my reasoning a bit. I was thinking about this with regards to `guides`, which was when I introduced the naming convention `record` (which may not be the right term — `resource` seems like the better way to go). As I was adapting some of the logic that loads `posts` to accommodate a different resource type (in my thinking there, `record`), I started thinking about `collections` as collections of records. So ideally, a `loadRecordCollections` function could ingest a `recordType` and check whether or not a collection (in the sense of a series) of that particular resource type exists, and if so, implement the right type checking and methods for that.  

In this way, then, I think it may make sense to use terminology that indicates that a collection could be a series of a particular resource type.

File: components/Article.tsx (hunk 8 above)
Comment (marekweb) [reply components/Article.tsx:29]: I see what you mean. Maybe if we ensure that we'll always use `null`, and not `undefined`, then we could make the fields non-optional.

File: util/loadAuthorCollections.ts [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+
+function returnAuthorCollection(authors: AuthorCollection[]): AuthorCollection[] {
+    let validatedAuthorCollections: AuthorCollection[] = []
+    authors.map(authorCollectionDefinition => {
+        const normalizedAuthorCollection = normalizeAuthorCollectionDefinition(authorCollectionDefinition)
+        validatedAuthorCollections = [ ...validatedAuthorCollections, normalizedAuthorCollection ]
+    })
+
+    return validatedAuthorCollections
+}
```
Comment (marekweb) [inline util/loadAuthorCollections.ts:20]: Since `map` returns the modified array you could simplify the function:
```ts
function returnAuthorCollection(authors: AuthorCollection[]): AuthorCollection[] {
    return authors.map(authorCollectionDefinition => normalizeAuthorCollectionDefinition(authorCollectionDefinition))
}
```

Or further:

```ts
function returnAuthorCollection(authors: AuthorCollection[]): AuthorCollection[] {
    return authors.map(normalizeAuthorCollectionDefinition)
}
```

File: util/validators.ts [hunk 11]
```diff
@@ -29,3 +35,37 @@ export function normalizeString(rawString: unknown): string {
...
+export function normalizeRecordCollectionDefinition(collection: RecordCollectionDefinition): RecordCollectionDefinition {
+    return {
+        title: normalizeString(collection.title),
+        slug: normalizeString(collection.slug),
+        type: normalizeString(collection.type),
+        members: normalizeStringArray(collection.members)
+    }
+}
+
+export function normalizeAuthorCollectionDefinition(author: AuthorCollection): AuthorCollection {
```
Comment (marekweb) [inline util/validators.ts:64]: The normalize functions look good to me. Just a thought for the future: If we normalize each field, I wonder if Typescript would let us define this as:
```suggestion
export function normalizeAuthorCollectionDefinition(author: unknown): AuthorCollection {
```
... which would let us drop the `as` clause when we load it in:

https://github.com/sourcegraph/learn/blob/f7335858acfb8f33d2c5e951742fbf6f7b0bb9cb/util/loadAuthorCollections.ts#L25

File: util/omitUndefinedFields.ts (hunk 9 above)
Comment (marekweb) [reply util/omitUndefinedFields.ts:17]: I think adding `|| value === null` here (above) might fix the null bug, returning early if we have a null value.
```ts
    if (typeof object !== 'object' || value === null) {
        return object
    }
```

File: util/loadAuthorCollections.ts (hunk 10 above)
Comment (katjuell) [reply util/loadAuthorCollections.ts:20]: Nice!

File: util/omitUndefinedFields.ts (hunk 9 above)
Comment (marekweb) [reply util/omitUndefinedFields.ts:17]: Sorry, I just realized my snippet is wrong, the variable is named `object`, not `value`

Comment (katjuell) [reply util/omitUndefinedFields.ts]: All good! I flipped the logic, which I thought may make sense. What do you think?

```
if (object && typeof object === 'object')
```
If it's not `null` and is an object then do the thing. Otherwise return the object.

File: components/Article.tsx (hunk 8 above)
Comment (ltagliaferri) [reply components/Article.tsx:29]: I think we should keep the fields as optional, I think the front matter can be a lot to ask for and in practice I see many contributors who work in markdown ignoring headers completely.

**Comment by ltagliaferri** [comment]
This looks good to me. We can revisit `records` vs `resources` vs `collections` (collections of authors, collections of records/resources).
